### PR TITLE
aws_vpn_connection: fix possible remote state data inconsistency

### LINKED ISSUE #396: VPN Connection Tunnels incorrectly ordered
_This issue was originally opened by @kitforbes as hashicorp/terraform#10411. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
Terraform v0.7.13

### Affected Resource(s)
- aws_vpn_connection

### Terraform Configuration Files
```hcl
module "vpc" {
  source = "../../../modules/aws/network/vpc"

  name        = "${var.name}-vpc"
  environment = "${var.environment}"
  cidr        = "${var.vpc_cidr}"
}

resource "aws_customer_gateway" "edinburgh" {
  type       = "${var.vpn_connection_type}"
  bgp_asn    = "${var.vpn_bgs_asn}"
  ip_address = "${var.vpn_local_ip}"

  tags {
    Name        = "${var.name}-customer-gateway"
    Environment = "${var.environment}"
    Creator     = "Terraform"
  }
}

resource "aws_vpn_gateway" "edinburgh" {
  vpc_id = "${module.vpc.vpc_id}"

  tags {
    Name        = "${var.name}-edinburgh"
    Environment = "${var.environment}"
    Creator     = "Terraform"
  }
}

resource "aws_vpn_connection" "edinburgh" {
  vpn_gateway_id      = "${aws_vpn_gateway.edinburgh.id}"
  customer_gateway_id = "${aws_customer_gateway.edinburgh.id}"
  type                = "${var.vpn_connection_type}"
  static_routes_only  = true

  tags {
    Name        = "${var.name}-edinburgh"
    Environment = "${var.environment}"
    Creator     = "Terraform"
  }
}

output "vpn_config" {
  value = <<CONFIGURATION

  Tunnel 1
    Address:     ${aws_vpn_connection.edinburgh.tunnel1_address}
  Tunnel 2
    Address:     ${aws_vpn_connection.edinburgh.tunnel2_address}
CONFIGURATION
}
```

### Expected Behavior
`${aws_vpn_connection.edinburgh.tunnel1_address}` should output the value of Tunnel 1 in the AWS Console.

### Actual Behavior
It actually outputs the value of tunnel 2, and tunnel 2 outputs the value of tunnel 1. Not a critical issue by any means.

### Steps to Reproduce
Create an AWS VPN connection and compare the Terraform output to actual AWS Console information in the VPC section.

### LINKED ISSUE #5809: aws_vpn_connection  - inside_tunnel_cidr swapped between tunnel interfaces
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
Terraform v0.11.8
+ provider.aws v1.35.0

### Affected Resource(s)

"aws_vpn_connection"

### Terraform Configuration Files

We create resource by using module 
```hcl
module "vpn_gateway_main" {
  source = "../terraform-aws-vpn-gateway.git"
  create_vpn_connection = "${ var.create_vpn  ? 1 : 0}"   

  vpn_gateway_id      = "${module.vpc.vgw_id}"
  customer_gateway_id = "${local.vpn_main_cgw_id}"

  tunnel1_inside_cidr   = "${local.vpn1_tunnel1_cidr}"
  tunnel2_inside_cidr   = "${local.vpn1_tunnel2_cidr}"
  vpc_id                       = "${module.vpc.vpc_id}"
  create_vpn_gateway_attachment = false 
}
```

from module resource is created 
https://github.com/terraform-aws-modules/terraform-aws-vpn-gateway/blob/master/main.tf
```
resource "aws_vpn_connection" "tunnel" {
  count = "${var.create_vpn_connection && length(var.tunnel1_inside_cidr) > 0 && length(var.tunnel2_inside_cidr) > 0 && length(var.tunnel1_preshared_key) == 0 && length(var.tunnel2_preshared_key) == 0 ? 1 : 0}"

  vpn_gateway_id      = "${var.vpn_gateway_id}"
  customer_gateway_id = "${var.customer_gateway_id}"
  type                = "ipsec.1"

  static_routes_only = "${var.vpn_connection_static_routes_only}"

  tunnel1_inside_cidr = "${var.tunnel1_inside_cidr}"
  tunnel2_inside_cidr = "${var.tunnel2_inside_cidr}"

  tags = "${merge(
    map(
      "Name", "VPN Connection between VPC ${var.vpc_id} and Customer Gateway ${var.customer_gateway_id}"
    ),
    var.tags
  )}"
}
```

### Debug Output

```
#terraform apply

module.account_vpc1.module.vpn_gateway_main.aws_vpn_connection.tunnel: Creating...
  customer_gateway_configuration: "" => "<computed>"
  customer_gateway_id:            "" => "cgw-07eb19a67ffa83222"
  routes.#:                       "" => "<computed>"
  static_routes_only:             "" => "false"
  tags.%:                         "" => "1"
  tags.Name:                      "" => "VPN Connection between VPC vpc-070021af585778888 and Customer Gateway cgw-07eb19a67ffa83222"
  tunnel1_address:                "" => "<computed>"
  tunnel1_bgp_asn:                "" => "<computed>"
  tunnel1_bgp_holdtime:           "" => "<computed>"
  tunnel1_cgw_inside_address:     "" => "<computed>"
  tunnel1_inside_cidr:            "" => "169.254.10.0/30"
  tunnel1_preshared_key:          "<sensitive>" => "<sensitive>"
  tunnel1_vgw_inside_address:     "" => "<computed>"
  tunnel2_address:                "" => "<computed>"
  tunnel2_bgp_asn:                "" => "<computed>"
  tunnel2_bgp_holdtime:           "" => "<computed>"
  tunnel2_cgw_inside_address:     "" => "<computed>"
  tunnel2_inside_cidr:            "" => "169.254.10.4/30"
  tunnel2_preshared_key:          "<sensitive>" => "<sensitive>"
  tunnel2_vgw_inside_address:     "" => "<computed>"
  type:                           "" => "ipsec.1"
  vgw_telemetry.#:                "" => "<computed>"
  vpn_gateway_id:                 "" => "vgw-024f5f47d937a4adb"

module.account_vpc1.module.vpn_gateway_secondary.aws_vpn_connection.tunnel: Creating...
  customer_gateway_configuration: "" => "<computed>"
  customer_gateway_id:            "" => "cgw-0f5f408a10dda83111"
  routes.#:                       "" => "<computed>"
  static_routes_only:             "" => "false"
  tags.%:                         "" => "1"
  tags.Name:                      "" => "VPN Connection between VPC vpc-070021af585778888 and Customer Gateway cgw-0f5f408a10dda83111"
  tunnel1_address:                "" => "<computed>"
  tunnel1_bgp_asn:                "" => "<computed>"
  tunnel1_bgp_holdtime:           "" => "<computed>"
  tunnel1_cgw_inside_address:     "" => "<computed>"
  tunnel1_inside_cidr:            "" => "169.254.10.8/30"
  tunnel1_preshared_key:          "<sensitive>" => "<sensitive>"
  tunnel1_vgw_inside_address:     "" => "<computed>"
  tunnel2_address:                "" => "<computed>"
  tunnel2_bgp_asn:                "" => "<computed>"
  tunnel2_bgp_holdtime:           "" => "<computed>"
  tunnel2_cgw_inside_address:     "" => "<computed>"
  tunnel2_inside_cidr:            "" => "169.254.10.12/30"
  tunnel2_preshared_key:          "<sensitive>" => "<sensitive>"
  tunnel2_vgw_inside_address:     "" => "<computed>"
  type:                           "" => "ipsec.1"
  vgw_telemetry.#:                "" => "<computed>"
  vpn_gateway_id:                 "" => "vgw-024f5f47d937a4adb"
```

### Panic Output

```
#terraform state show module.account_vpc1.module.vpn_gateway_main.aws_vpn_connection.tunnel

VPN connection 1 
customer_gateway_id                           = cgw-07eb19a67ffa83222
routes.#                                      = 0
static_routes_only                            = false
tags.%                                        = 1
tags.Name                                     = VPN Connection between VPC vpc-070021af58578888 and Customer Gateway cgw-07eb19a67ffa83222
tunnel1_address                               = 18.184.98.78
tunnel1_bgp_asn                               = 64512
tunnel1_bgp_holdtime                          = 30
tunnel1_cgw_inside_address                    = 169.254.10.6
tunnel1_inside_cidr                           = 169.254.10.0/30
tunnel1_preshared_key                         = password
tunnel1_vgw_inside_address                    = 169.254.10.5
tunnel2_address                               = 35.158.222.155
tunnel2_bgp_asn                               = 64512
tunnel2_bgp_holdtime                          = 30
tunnel2_cgw_inside_address                    = 169.254.10.2
tunnel2_inside_cidr                           = 169.254.10.4/30
tunnel2_preshared_key                         = password
tunnel2_vgw_inside_address                    = 169.254.10.1

#terraform state show module.account_vpc1.module.vpn_gateway_secondary.aws_vpn_connection.tunnel

customer_gateway_id                           = cgw-0f5f408a10dda8111
routes.#                                      = 0
static_routes_only                            = false
tags.%                                        = 1
tags.Name                                     = VPN Connection between VPC vpc-070021af585778888 and Customer Gateway cgw-0f5f408a10dda8111
tunnel1_address                               = 18.196.49.49
tunnel1_bgp_asn                               = 64512
tunnel1_bgp_holdtime                          = 30
tunnel1_cgw_inside_address                    = 169.254.10.10
tunnel1_inside_cidr                           = 169.254.10.8/30
tunnel1_preshared_key                         = password
tunnel1_vgw_inside_address                    = 169.254.10.9
tunnel2_address                               = 35.158.82.179
tunnel2_bgp_asn                               = 64512
tunnel2_bgp_holdtime                          = 30
tunnel2_cgw_inside_address                    = 169.254.10.14
tunnel2_inside_cidr                           = 169.254.10.12/30
tunnel2_preshared_key                         = password
tunnel2_vgw_inside_address                    = 169.254.10.13
type                                          = ipsec.1
```

### Expected Behavior

We created two VPN connections called in output "main" and "secondary"  to separate Customer gateways. Both IPSEC Connection have inside two tunnels.
To prevent conflicts we manually assign CIDR inside each tunnel.

First ipsec connection has
Connection "main":
tunnel1_cidr: 169.254.10.0/30
tunnel2_cidr: 169.254.10.4/30

Connection "secondary":
tunnel1_cidr: 169.254.8.0/30
tunnel2_cidr: 169.254.12.4/30

### Actual Behavior

From time to time I can see that IP addresses are swapped between tunnel interfaces within a single connection.
It doesn't occur each time but quite often 

As we can see in terraform state show command we have the problem in first connection.
We assigned CIDR 169.254.10.0/30 and is visible while we are creating resource but we at the end obtain addresses for second interfaces 

tunnel1_cgw_inside_address                    = 169.254.10.6
tunnel1_inside_cidr                           = 169.254.10.0/30
tunnel1_vgw_inside_address                    = 169.254.10.5

tunnel2_cgw_inside_address                    = 169.254.10.2
tunnel2_inside_cidr                           = 169.254.10.4/30
tunnel2_vgw_inside_address                    = 169.254.10.1
 

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

### LINKED ISSUE #11293: aws_vpn_connection: inconsistent remote state data (occasionally)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.12.17
+ provider.aws v2.42.0

### Affected Resource(s)

* aws_vpn_connection

### Terraform Configuration Files

**does not matter**

### Background

`resourceAwsVpnConnectionCreate` assembles the connection options [1] and calls AWS API to create the VPN connection with these accordingly. It seems AWS is creating the VPN connections asynchronously as sometimes the order of the VPN connections in the XML config blob does not apply to the order of `options[]`. Since `tunnel1_inside_cidr` and `tunnel2_inside_cidr` are saved directly to the remote state and all other attributes are added later in `resourceAwsVpnConnectionRead` [2], the data for the resource might be inconsistent.

The code already tries to cover this unsuccessfully via sorting the XML config blob [3]. `sort.Sort(vpnConfig)` is acting on `OutsideAddress` of `type XmlIpsecTunnel struct` [4].

We need to check wether:
- `vpnConfig.Tunnels[0].VgwInsideAddress` or `vpnConfig.Tunnels[0].CgwInsideAddress` is a vaild address of `tunnel1_inside_cidr`
- `vpnConfig.Tunnels[1].VgwInsideAddress` or `vpnConfig.Tunnels[1].CgwInsideAddress` is a vaild address of `tunnel2_inside_cidr`

[1] https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_vpn_connection.go#L277-L291
[2] https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_vpn_connection.go#L452-L463
[3] https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_vpn_connection.go#L596
[4] https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_vpn_connection.go#L27-L28

### Expected Behavior

```console
terraform state show aws_vpn_connection.example | grep -e tunnel1 -e tunnel2 | grep -v -e tunnel1_address -e tunnel2_address  -e bgp -e key
    tunnel1_cgw_inside_address     = "169.254.254.250"
    tunnel1_inside_cidr            = "169.254.254.248/30"
    tunnel1_vgw_inside_address     = "169.254.254.249"
    tunnel2_cgw_inside_address     = "169.254.254.254"
    tunnel2_inside_cidr            = "169.254.254.252/30"
    tunnel2_vgw_inside_address     = "169.254.254.253"
```

### Actual Behavior

```console
terraform state show aws_vpn_connection.example | grep -e tunnel1 -e tunnel2 | grep -v -e tunnel1_address -e tunnel2_address  -e bgp -e key
    tunnel1_cgw_inside_address     = "169.254.254.250"
    tunnel1_inside_cidr            = "169.254.254.252/30"
    tunnel1_vgw_inside_address     = "169.254.254.249"
    tunnel2_cgw_inside_address     = "169.254.254.254"
    tunnel2_inside_cidr            = "169.254.254.248/30"
    tunnel2_vgw_inside_address     = "169.254.254.253"
```

### References
- #396
- #5809

### PR DESCRIPTION
**Pull request by n3ph**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11293
Closes #396
Closes #5809
Closes #10584

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
aws_vpn_connection: fix possible remote state data inconsistency 
```

Output from acceptance testing:

After substantiation of test case (https://github.com/terraform-providers/terraform-provider-aws/pull/11298/commits/20b9373b1075e26f4e8461e480c918dfcc7187b2):
```console
make testacc TEST=./aws TESTARGS='-run=TestAWSVpnConnection_xmlconfig'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAWSVpnConnection_xmlconfig -timeout 120m
=== RUN   TestAWSVpnConnection_xmlconfig
--- FAIL: TestAWSVpnConnection_xmlconfig (0.00s)
    resource_aws_vpn_connection_test.go:397: First address from tunnel XML was incorrect.
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	0.019s
FAIL
make: *** [GNUmakefile:24: testacc] Error 1
```

After fixing the sorting (https://github.com/terraform-providers/terraform-provider-aws/pull/11298/commits/ac712d126081f3e8ae60ec593ef60e8904959cff):
```console
 make testacc TEST=./aws TESTARGS='-run=TestAWSVpnConnection_xmlconfig'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAWSVpnConnection_xmlconfig -timeout 120m
=== RUN   TestAWSVpnConnection_xmlconfig
--- PASS: TestAWSVpnConnection_xmlconfig (0.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	0.018s
```

### CODE REVIEW

File: aws/resource_aws_vpn_connection.go [hunk 1]
```diff
@@ -586,14 +587,20 @@ func waitForEc2VpnConnectionDeletion(conn *ec2.EC2, id string) error {
...
 	var vpnConfig XmlVpnConnectionConfig
 	if err := xml.Unmarshal([]byte(xmlConfig), &vpnConfig); err != nil {
 		return nil, fmt.Errorf("Error Unmarshalling XML: %s", err)
 	}
 
-	// don't expect consistent ordering from the XML
-	sort.Sort(vpnConfig)
+	// don't expect consistent ordering
+	_, ipnet1, _ := net.ParseCIDR(tunnel1_inside_cidr)
+	if !(ipnet1.Contains(net.ParseIP(vpnConfig.Tunnels[0].VgwInsideAddress))) {
```
Comment (bflad) [inline aws/resource_aws_vpn_connection.go]: Please note that this line currently can panic:

```
=== CONT  TestAccAWSVpnConnection_basic

------- Stderr: -------
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x615726]

goroutine 358 [running]:
net.networkNumberAndMask(0x0, 0xc15bfea9a9fe5bc1, 0x3, 0xc00158e660, 0xc000adb270, 0x616539, 0xc00158e2ab)
	/opt/goenv/versions/1.13.0/src/net/ip.go:473 +0x26
net.(*IPNet).Contains(0x0, 0xc00158e660, 0x10, 0x10, 0x10)
	/opt/goenv/versions/1.13.0/src/net/ip.go:497 +0x2f
github.com/terraform-providers/terraform-provider-aws/aws.xmlConfigToTunnelInfo(0xc000e81000, 0xe73, 0x0, 0x0, 0xa1a56e0, 0x0, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_vpn_connection.go:598 +0x1c3
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsVpnConnectionRead(0xc0000fd0a0, 0x4c5fe80, 0xc0000ec000, 0x0, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_vpn_connection.go:450 +0xe30
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsVpnConnectionCreate(0xc0000fd0a0, 0x4c5fe80, 0xc0000ec000, 0x2, 0xa1a56e0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_vpn_connection.go:330 +0x7fc
```

Comment (n3ph) [reply aws/resource_aws_vpn_connection.go]: Not sure how to elegantly catch that.. Any hints?

**Comment by n3ph** [comment]
... forgot `--rebase` in `git pull` :-/

File: aws/resource_aws_vpn_connection.go (hunk 1 above)
Comment (n3ph) [reply aws/resource_aws_vpn_connection.go]: Right - Got it. `tunnel1_inside_cidr` is optional. So checking for nil should do the trick..

**Comment by n3ph** [comment]
Fucked up the rebase - Anyhow - nobody seemed to be interested

**Pull request closed without merging**